build-on-push/platform_ci/platform_ci/jenkins.py

```diff
@@ -0,0 +1,229 @@
+        subclass (this allows to switch to a different Jenkins API.
+        """
+        return PlatformJenkinsJavaCLI(template_dir, url)
+
+    def __init__(self, jenkins, template_dir):
+        self.jenkins_server = jenkins
+        self.template_dir = template_dir
+
+
+class PlatformJenkinsJavaCLI(PlatformJenkins):
+    """Represents a Platform CI Jenkins instance, wrapping a Java CLI API.
+
+    While ugly to wrap the Java CLI in Python like this, we had the best
+    experience with using the Java CLI from all other Jenkins API options
+    that we tried, especially related to authentication.
+
+    If someone implements a replacement class over Python Jenkins API that
+    *works*, it would probably be best to replace this class with it.
+    """
+
+    CLI = ["/usr/bin/java", "-jar", "/var/lib/jenkins/jenkins-cli.jar", "-noCertificateCheck"]
+
+    GET_VIEW = "get-view"
+    UPDATE_VIEW = "update-view"
+    CREATE_VIEW = "create-view"
+
+    VIEW_TEMPLATE = "view-template.xml"
+
+    LIST_JOBS = "list-jobs"
+    DELETE_JOB = "delete-job"
+    BUILD_JOB = "build"
+    CREATE_JOB = "create-job"
+    UPDATE_JOB = "update-job"
+    ENABLE_JOB = "enable-job"
+    DISABLE_JOB = "disable-job"
+
+    SET_DESCRIPTION = "set-build-description"
+
+    def __init__(self, template_dir, url):
+        super(PlatformJenkinsJavaCLI, self).__init__(None, template_dir)
+        self.url = url
+        self.cli = []
+        if "BOP_JENKINS_CLI" in os.environ:
+            self.cli.extend(os.environ["BOP_JENKINS_CLI"].split())
+            self.cli.append("-noCertificateCheck")
+        else:
+            self.cli.extend(PlatformJenkinsJavaCLI.CLI)
+
+        self.cli.extend(["-s", url])
+
+    def view_exists(self, view):
+        """Returns true if a given view exists."""
+        with open("/dev/null", "w") as devnull:
+            call = subprocess.Popen(self.cli + [PlatformJenkinsJavaCLI.GET_VIEW, view], stdout=devnull, stderr=devnull)
+            call.wait()
+        return call.returncode == 0
+
+    def set_view(self, view, view_xml_filename):
+        """Creates a View, defined by XML in view_xml_filename.
+
+        If the file exists, it will be update using the provided definition.
+
+        Args:
+            view: Created view name
+            view_xml_filename: Path to a file containing a XML definition of a view.
+        """
+        if self.view_exists(view):
+            command = PlatformJenkinsJavaCLI.UPDATE_VIEW
+        else:
+            command = PlatformJenkinsJavaCLI.CREATE_VIEW
+
+        with open(view_xml_filename) as view_xml_file:
+            view_xml = view_xml_file.read()
+
+        call = subprocess.Popen(self.cli + [command, view], stdin=subprocess.PIPE)
+        call.communicate(view_xml)
+        call.wait()
+
+    def job_exists(self, job):
+        """Returns True if the given job exists."""
+        call = subprocess.Popen(self.cli + [PlatformJenkinsJavaCLI.LIST_JOBS], stdout=subprocess.PIPE)
```

Comment:
IMO we need a better way to check if job_exists. 'list-jobs' when there are thousands of jobs doesn't seem the right approach.